### PR TITLE
reactor_backend: Fix another busy spin bug in the epoll backend

### PR DESCRIPTION
**Pull request by StephanDollberg**
Previously we fixed a busy spinning bug in the epoll reactor backend
in https://github.com/scylladb/seastar/pull/1945 / https://github.com/scylladb/seastar/commit/25e2487da55fb37d39ecd72e68c8e9d87f784c62.

The original scenario in that case was very easy to reproduce by just
deferring the socket close on error.

The root cause is the delayed removal of the fd from the epoll set (this
is an optimization I presume to avoid unnecessary syscalls) together
with how the logic is implemented.

Now we have seen similar behaviour in the wild and I found at least one
other race condition that can lead to the same spinning symptom.

To following sequence of events can lead to busy spinning:

 - We are trying to read data and hence are waiting for EPOLLIN
 - EPOLLIN comes in. We process the data and remove EPOLLIN from
   `pfd->events_requested` which subsequently gets set to 0. Importantly
   we are not immediately trying to read from the connection again.
   Maybe the connection got put back onto a connection pool or similar.
 - Because of the optimization explained above in this poll tick we are
   not yet going to remove the fd from the epoll set (this would only
   happen on the next tick).
 - The next tick comes in but in the meantime we got a RST and hence we
   get EPOLLIN | EPOLLERR instead.
 - Now the has_error branch triggers from our previous patch and
   events_to_remove gets resolved to `pfd->events_requested` which is 0
   and hence the fd won't get removed from the epoll set
 - This busy repeats until we close the socket or try to read from it
   again.

How easy it is to trigger this in real life depends on a few factors but
it's easy to reproduce by inducing a slow tick (maybe via a reactor
stall) during which the RST comes in.

To avoid this issue we are now adding a more defensive branch which just
unconditionally removes the fd from the epoll set if we have errors
(note getting an error means that we have already triggered completition
of all outstanding events).

If you look at the original review linked above you'll find that that is
what I had suggested originally but we then moved away from it given
review comments. I think having this more defensive approach is the
safest thing here given the low test coverage of the epoll backend.

### CODE REVIEW

File: src/core/reactor_backend.cc [hunk 1]
```diff
@@ -873,7 +873,7 @@ reactor_backend_epoll::wait_and_process(int timeout, const sigset_t* active_sigm
             evt.events = pfd->events_requested;
         }
         auto events = evt.events & (EPOLLIN | EPOLLOUT | EPOLLRDHUP);
-        auto events_to_remove = has_error ? pfd->events_requested : events & ~pfd->events_requested;
+        auto events_to_remove = events & ~pfd->events_requested;
```
Comment (StephanDollberg) [inline src/core/reactor_backend.cc:876]: Note this line effectively just "reverts" the previous commit as the added code below handles both cases (the original bug plus the race condition described in this commit plus possibly other unknown ones).

**Comment by avikivity** [comment]
Can you add a unit test?

**Comment by StephanDollberg** [comment]
I'd love to but it's tricky to do because these bugs have no external visiblity other than the busy spinning. We could check the metric for that. 

A test for the race condition described above would probably be too tricky to make reliably work but I could give it a try for the original issue which is easier to reproduce. I am just worried about flakiness.

I can give it a try.

**Comment by avikivity** [comment]
Ah, it doesn't lock up, just avoids sleeping?

We do have metrics for sleeping etc.

**Comment by StephanDollberg** [comment]
> Ah, it doesn't lock up, just avoids sleeping?

Yeah exactly, it's just a (benign) busy spin. Reactor util will show as 100% which for the end user can obviously look alerting.

In any case, I have added a unit test for the original scenario.

File: tests/unit/epoll_test.cc [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2026 ScyllaDB
```
Comment (avikivity) [inline tests/unit/epoll_test.cc]: Misattribution

**Comment by avikivity** [comment]
The test has many false-negatives when run without the fix:

```
$ build/dev/tests/unit/epoll_test --  --reactor-backend epoll
Running 1 test case...
INFO  2026-01-15 20:42:16,017 seastar - Reactor backend: epoll
INFO  2026-01-15 20:42:16,093 [shard 16:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
INFO  2026-01-15 20:42:16,094 [shard 16:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
WARN  2026-01-15 20:42:16,150 seastar - Unable to set SCHED_FIFO scheduling policy for timer thread; latency impact possible. Try adding CAP_SYS_NICE
random-seed=1451047005

*** No errors detected
$ build/dev/tests/unit/epoll_test --  --reactor-backend epoll
Running 1 test case...
INFO  2026-01-15 20:42:26,638 seastar - Reactor backend: epoll
INFO  2026-01-15 20:42:26,729 [shard 16:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
INFO  2026-01-15 20:42:26,729 [shard 16:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
WARN  2026-01-15 20:42:26,760 seastar - Unable to set SCHED_FIFO scheduling policy for timer thread; latency impact possible. Try adding CAP_SYS_NICE
random-seed=3771538806

*** No errors detected
$ build/dev/tests/unit/epoll_test --  --reactor-backend epoll
Running 1 test case...
INFO  2026-01-15 20:42:30,816 seastar - Reactor backend: epoll
INFO  2026-01-15 20:42:30,929 [shard  4:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
INFO  2026-01-15 20:42:30,929 [shard  4:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
WARN  2026-01-15 20:42:30,942 seastar - Unable to set SCHED_FIFO scheduling policy for timer thread; latency impact possible. Try adding CAP_SYS_NICE
random-seed=2889462350

*** No errors detected
$ build/dev/tests/unit/epoll_test --  --reactor-backend epoll
Running 1 test case...
INFO  2026-01-15 20:42:35,647 seastar - Reactor backend: epoll
INFO  2026-01-15 20:42:35,714 [shard 16:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
INFO  2026-01-15 20:42:35,715 [shard 16:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
WARN  2026-01-15 20:42:35,782 seastar - Unable to set SCHED_FIFO scheduling policy for timer thread; latency impact possible. Try adding CAP_SYS_NICE
random-seed=3869718020

*** No errors detected
$ time build/dev/tests/unit/epoll_test --  --reactor-backend epoll
Running 1 test case...
INFO  2026-01-15 20:42:52,210 seastar - Reactor backend: epoll
INFO  2026-01-15 20:42:52,306 [shard 17:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
INFO  2026-01-15 20:42:52,306 [shard 17:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
WARN  2026-01-15 20:42:52,334 seastar - Unable to set SCHED_FIFO scheduling policy for timer thread; latency impact possible. Try adding CAP_SYS_NICE
random-seed=3456995436

*** No errors detected

real	0m3.398s
user	0m1.505s
sys	0m2.348s
$ time build/dev/tests/unit/epoll_test --  --reactor-backend epoll --smp 2
Running 1 test case...
INFO  2026-01-15 20:43:10,468 seastar - Reactor backend: epoll
WARN  2026-01-15 20:43:10,479 seastar - Unable to set SCHED_FIFO scheduling policy for timer thread; latency impact possible. Try adding CAP_SYS_NICE
INFO  2026-01-15 20:43:10,479 [shard 0:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
INFO  2026-01-15 20:43:10,479 [shard 0:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
random-seed=2708152722

*** No errors detected

real	0m3.259s
user	0m0.059s
sys	0m0.093s
$ time build/dev/tests/unit/epoll_test --  --reactor-backend epoll --smp 2
Running 1 test case...
INFO  2026-01-15 20:43:14,742 seastar - Reactor backend: epoll
WARN  2026-01-15 20:43:14,760 seastar - Unable to set SCHED_FIFO scheduling policy for timer thread; latency impact possible. Try adding CAP_SYS_NICE
INFO  2026-01-15 20:43:14,760 [shard 0:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
INFO  2026-01-15 20:43:14,760 [shard 0:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
random-seed=302471684
/home/avi/seastar/tests/unit/epoll_test.cc(85): fatal error: in "epoll_busy_spin_on_socket_error_test": critical check busy_ms < 1500 has failed [3000 >= 1500]

*** 1 failure is detected in the test module "Master Test Suite"

real	0m3.285s
user	0m1.647s
sys	0m1.477s
$ time build/dev/tests/unit/epoll_test --  --reactor-backend epoll --smp 2
Running 1 test case...
INFO  2026-01-15 20:43:20,436 seastar - Reactor backend: epoll
WARN  2026-01-15 20:43:20,450 seastar - Unable to set SCHED_FIFO scheduling policy for timer thread; latency impact possible. Try adding CAP_SYS_NICE
INFO  2026-01-15 20:43:20,450 [shard 0:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
INFO  2026-01-15 20:43:20,450 [shard 0:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
random-seed=2368374268

*** No errors detected

real	0m3.238s
user	0m0.069s
sys	0m0.083s
$ time build/dev/tests/unit/epoll_test --  --reactor-backend epoll --smp 2
Running 1 test case...
INFO  2026-01-15 20:43:27,826 seastar - Reactor backend: epoll
WARN  2026-01-15 20:43:27,843 seastar - Unable to set SCHED_FIFO scheduling policy for timer thread; latency impact possible. Try adding CAP_SYS_NICE
INFO  2026-01-15 20:43:27,843 [shard 0:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
INFO  2026-01-15 20:43:27,843 [shard 0:main] io - IO queue was unable to find a suitable maximum request length, the search was cut-off early at: 16MB
random-seed=1424443070
/home/avi/seastar/tests/unit/epoll_test.cc(85): fatal error: in "epoll_busy_spin_on_socket_error_test": critical check busy_ms < 1500 has failed [3000 >= 1500]

*** 1 failure is detected in the test module "Master Test Suite"

real	0m3.277s
user	0m1.670s
sys	0m1.451s
```

File: tests/unit/epoll_test.cc (hunk 2 above)
Comment (StephanDollberg) [reply tests/unit/epoll_test.cc]: Just copied from some other test file. Whom do you want me to attribute it to?

**Comment by StephanDollberg** [comment]
Yes because that is the "easy" case which is already fixed by my previous commit which has been on master for a while. You need to also revert https://github.com/scylladb/seastar/commit/25e2487da55fb37d39ecd72e68c8e9d87f784c62 or just comment out the extra "else if" in the version of this PR (which makes it equivalent to pre 25e2487da55fb37d39ecd72e68c8e9d87f784c62).

The fact you got one failure is actually interesting and shows that the original patch was not complete.

File: tests/unit/epoll_test.cc (hunk 2 above)
Comment (avikivity) [reply tests/unit/epoll_test.cc]: Whoever wrote it or their employer, according to their policy

**Comment by avikivity** [comment]
Ok. It's generally better not to have flaky tests (in either the false-positive or false-negative direction) but this is such an edge case I'll accept the current test.

File: tests/unit/epoll_test.cc (hunk 2 above)
Comment (StephanDollberg) [reply tests/unit/epoll_test.cc]: Fixed

**Comment by StephanDollberg** [comment]
ping @avikivity
